File: keepr-client/src/services/KeeprsService.js
import { api } from './AxiosService'
import { AppState } from '../AppState'
import { logger } from '../utils/Logger'
import { Keepr } from '../Models/KeeprModel'

export default class KeeprsService {
  async get() {
    try {
      const res = await api.get('api/keeps')
      logger.log('ANything', res.data)
      AppState.keeps = res.data.map(k => new Keepr(k))
    } catch (error) {
      logger.error('This is the error', error)
    }
  }

  async getById(id) {
    try {
      const res = await api.get('api/keeps/' + id)
      AppState.keepr = res.data
    } catch (error) {
      logger.error(error)
    }
  }

  async getKeepsByVaultId(id) {
    try {
      const res = await api.get('api/vaults/' + id + '/keeps')
      AppState.keepr.vaults = res.data
      AppState.vault = res.data
      logger.log(res.data)
    } catch (error) {
      logger.error(error)
    }
  }

  async getKeepsByProfileId(id) {
    try {
      const res = await api.get('profile/' + id + '/keeps')
      AppState.keepr.profile = res.data
      AppState.profile = res.data
      logger.log(res.data)
    } catch (error) {
      logger.error(error)
    }
  }

  async createKeepr(keepr) {
    try {
      const res = await api.post('api/keeps/', keepr)
      AppState.keepr.push(res.data)
      return res.data
    } catch (error) {
      logger.error(error)
    }
  }

  async edit(keepr) {
    try {
      const res = await api.put('api/keeps/' + keepr.id, keepr)
      this.getById(res.data._id)
    } catch (error) {
      logger.error(error)
    }
  }

  async delete(id) {
    const res = window.confirm('Are you sure you want to delete this Keepr')
    if (!res) {
      return
    }
    try {
      await api.delete('api/keeps/' + id)
      this.getById(id)
    } catch (error) {
      logger.error(error)
    }
  }
}

export const keeprsService = new KeeprsService()

File: keepr-server/Services/KeeprsService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using keeprapp.Models;
using keeprapp.Repositories;

namespace keeprapp.Services
{
    public class KeeprsService
    {
        private readonly KeeprsRepository _repo;
        private readonly VaultsRepository _vrepo;
        public KeeprsService(KeeprsRepository repo, VaultsRepository vrepo)
        {
            _repo = repo;
            _vrepo = vrepo;

        }

        public IEnumerable<Keepr> GetAll()
        {
            IEnumerable<Keepr> Keeps = _repo.GetAll();
            return Keeps;
        }

        internal Keepr GetById(int id)
        {
            var data = _repo.GetById(id);
            if (data == null)
            {
                throw new Exception("Invalid Id");
            }
            return data;
        }


        public Keepr Create(Keepr newKeepr)
        {
            newKeepr.Id = _repo.Create(newKeepr);
            return newKeepr;
        }

        internal Keepr Edit(Keepr editData, string userId)
        {
            Keepr original = GetById(editData.Id);
            if (original.CreatorId != userId)
            {
                throw new Exception("Access Denied: Cannot Edit a Keep You did not Create");

            }
            editData.Name = editData.Name == null ? original.Name : editData.Name;
            editData.Description = editData.Description == null ? original.Description : editData.Description;
            editData.Img = editData.Img == null ? original.Img : editData.Img;

            return _repo.Edit(editData);

        }

        internal IEnumerable<VaultKeepViewModel> GetKeepsByVaultId(int vaultId, Profile userInfo)
        {
            Vault original = _vrepo.Get(vaultId);
            if (original == null)
            {
                throw new Exception("Invalid Id");
            }
            if (original.IsPrivate == true && original.CreatorId != userInfo?.Id)
            {
                throw new Exception("This is private and you're not the owner!");
            }
            return _repo.GetKeepsByVaultId(vaultId);

        }

        internal string Delete(int id, string userId)
        {
            Keepr original = GetById(id);
            if (original.CreatorId != userId) 
            { throw new Exception("Access Denied: Cannot Delete a Keep You did not Create"); }
            _repo.Remove(id);
            return "successfully deleted";
        }

        internal IEnumerable<VaultKeepViewModel> GetKeepsByProfileId(string profileId)
        {
            IEnumerable<VaultKeepViewModel> keeps = _repo.GetKeepsByProfileId(profileId);
            return keeps.ToList();
        }


        internal IEnumerable<VaultKeepViewModel> GetByAccountId(string id)
        {
            return _repo.GetKeepsByProfileId(id);
        }



    }
}